libyara/include/yara/scanner.h

```diff
@@ -45,6 +48,10 @@ YR_API void yr_scanner_set_callback(
     YR_CALLBACK_FUNC callback,
     void* user_data);
 
+YR_API void yr_scanner_set_user_data_iterator(YR_SCANNER* scanner, void* user_data_iterator);
```

Comment:
I think that you don't really need to add these two new APIs. You want this because you want to pass your own information associated to the iterator, but `YR_MEMORY_BLOCK_ITERATOR` already have `context` field, which is a void pointer that functions `first` and `next` can use for whatever they want. In `_yr_test_single_or_multi_block_scan_mem` you do this:

```
 udi->iterator->context = udi->context;
```

Where `udi` is a pointer passed as `user_data_iterator`. It should be the other way around, the context that you assign to the iterator should be a structure that contains all the information needed for the `first` and `next` function to be able to run.

The `yr_scanner_scan_mem_blocks` function receives a `YR_MEMORY_BLOCK_ITERATOR` which for which you are going to provide your own `first` and `next` functions, these two functions receive a pointer to the `YR_MEMORY_BLOCK_ITERATOR` structure, and the `context` field in that structure is completely owned by you, you can assign whatever you want to it before calling `yr_scanner_scan_mem_blocks`, and you have access to that `context` from `first` and `next` . So, why are you needing this new `user_data_iterator` pointer?

Reply:
Okay, I removed `user_data_iterator` and also simplified the test iterator so that the single block and multi block iterators use the same data structure. `test-rules.c` also does a pass with the default iterator so it's okay that the default and single block test iterator use slightly different structures because they both get tested.

Reply:
@plusvic looks like Travis CI is doing that unrelated flappy build fail again?